File: steps/step05_skills.java
///usr/bin/env jbang "$0" "$@" ; exit $?
//JAVA 25
//DEPS com.fasterxml.jackson.core:jackson-databind:2.17.2
//SOURCES InnerHarness.java

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.Map;

// Step 5: Skills — feedforward, inferential, progressive disclosure.
// Same 2×2 cell as Step 4, different mechanism.
// "Skills for progressive disclosure of knowledge." — HumanLayer

void main(String[] args) {
    IO.println("=== Step 5: Skills (Feedforward, Inferential, Ollama) ===");

    // Load AGENTS.md
    var agentsMd = readFile("outer-harness/AGENTS.md");
    var agent = new InnerHarness.Agent(25);
    agent.prependSystemPrompt("# Project Conventions\n\n" + agentsMd);

    // Register the load_skill tool
    agent.registerTool(LoadSkill.SPEC, tu -> tu.input(LoadSkill.class).execute());

    var task = args.length > 0 ? String.join(" ", args)
            : "Write a test for LoanService.returnBook that verifies returning an overdue book still works correctly. Use the /how-to-write-tests skill for conventions.";

    agent.run(task);
}

static class LoadSkill {
    static final InnerHarness.ToolSpec SPEC = new InnerHarness.ToolSpec(
            "load_skill",
            "Load a skill by name. Skills provide detailed conventions and instructions. Available skills: how-to-write-tests, code-review, architecture-review. Call this when you need detailed guidance on a topic.",
            InnerHarness.objectSchema(Map.of(
                    "name", InnerHarness.stringProperty("The skill name, e.g. 'how-to-write-tests'")),
                    List.of("name")));

    public String name;

    public String execute() {
        var skillPath = Path.of("outer-harness/skills/" + name + ".md");
        try {
            var content = Files.readString(skillPath);
            return "=== Skill: " + name + " ===\n\n" + content;
        } catch (IOException e) {
            return "Skill not found: " + name + ". Available: how-to-write-tests, code-review, architecture-review";
        }
    }
}

String readFile(String path) {
    try { return Files.readString(Path.of(path)); }
    catch (IOException e) { return ""; }
}

File: python/steps/step05_skills.py
#!/usr/bin/env python3
"""Step 5: Skills -- feedforward, inferential, progressive disclosure.
Same 2x2 cell as Step 4, different mechanism.
'Skills for progressive disclosure of knowledge.' -- HumanLayer"""

import sys

from inner_harness import Agent, string_property


def load_skill(arguments):
    name = arguments.get("name", "")
    if not name:
        return "Error: name is required."
    try:
        with open(f"outer_harness/skills/{name}.md") as f:
            content = f.read()
        return f"=== Skill: {name} ===\n\n" + content
    except FileNotFoundError:
        return (
            f"Skill not found: {name}. "
            "Available: how-to-write-tests, code-review, architecture-review"
        )


def main():
    print("=== Step 5: Skills (Feedforward, Inferential, Ollama) ===")

    # Load AGENTS.md
    try:
        with open("outer_harness/AGENTS.md") as f:
            agents_md = f.read()
    except OSError:
        agents_md = ""

    agent = Agent(max_turns=25)
    agent.prepend_system_prompt("# Project Conventions\n\n" + agents_md)

    # Register the load_skill tool
    agent.register_tool(
        "load_skill",
        "Load a skill by name. Skills provide detailed conventions and instructions. "
        "Available skills: how-to-write-tests, code-review, architecture-review. "
        "Call this when you need detailed guidance on a topic.",
        {"name": string_property("The skill name, e.g. 'how-to-write-tests'")},
        load_skill,
    )

    task = (
        " ".join(sys.argv[1:])
        if len(sys.argv) > 1
        else "Write a test for LoanService.return_book that verifies returning "
             "an overdue book still works correctly. Use the /how-to-write-tests "
             "skill for conventions."
    )

    agent.run(task)


if __name__ == "__main__":
    main()
